### DISCUSSION TITLE
Use on Linux

### DISCUSSION THREAD

**Discussion by axgrinder** [pymodbus-dev/pymodbus#2447] [Q&A] [answered]
This is not a pymodbus issue... maybe. 

I have a some code based on and using this library. I can execute the code on windows and everything is perfect. When attempting to use on ubuntu 22.04, I can see using netstat -t the server is established. Using Modbus poll on another pc, 

I can connect but get timeout errors. 

Same issue using mbpoll on the same linux machine

Error the linux system running the server shows:

".../async_io.py, line 69, in callback_connected 
self.framer = self.server.framer(
Type error: 'str' object is not callable"

This is the exact same code that is flawless on windows. What am I missing? is it some permissions to allow the data out of the linux machine?

**Comment by janiversen** [comment]
Without seeing the code it's impossible to tell what the problem is.

Are you sure both systems use v3.7.4 the error looks like an old server.

**Comment by axgrinder** [comment]
Thank you for your response. Again, the codes works perfectly on windows 11 both with python3 and pycharm. Only on linux do I have an issue. here are the 2 scripts:
Server.py:
import asyncio

from pymodbus.datastore import (
    ModbusServerContext,
    ModbusSlaveContext,
    ModbusSparseDataBlock,
)
from pymodbus.server import (
    StartAsyncTcpServer,
)

import constants as c
import helper


def setup_server(context=None):
    """Run server setup."""
    args = helper.get_commandline(server=True)
    if context:
        args.context = context
    if not args.context:
        # setup registers FOR TOP LEVEL INFORMATION
        network_info = ModbusSparseDataBlock(c.NETWORK_REGS, mutable=True)
        system_info = ModbusSparseDataBlock(c.SYSTEM_REGS, mutable=True)

        #  Set the slave ids
        context = {
            0x01: ModbusSlaveContext(hr=network_info, ir=system_info)
        }

        # Build data storage
        args.context = ModbusServerContext(slaves=context, single=False)

    network_info = ModbusSparseDataBlock(c.NETWORK_REGS)
    system_info = ModbusSparseDataBlock(c.SYSTEM_REGS)

    context = {
        0x01: ModbusSlaveContext(hr=network_info, ir=system_info)
    }
    args.context = ModbusServerContext(slaves=context, single=False)
    return args


async def run_async_server(args):
    """Run server."""
    address = (args.host if args.host else "", args.port if args.port else None)
    server = await StartAsyncTcpServer(
        context=args.context,  # Data storage
        address=address,  # listen address
        framer=args.framer,  # The framer strategy to use
        ignore_missing_slaves=True,  # ignore request to a missing slave
    )

    return server


async def async_helper():
    """Combine setup and run."""
    run_args = setup_server()
    await run_async_server(run_args)


if __name__ == "__main__":
    asyncio.run(async_helper(), debug=False)


updater.py:
#!/usr/bin/env python3

import asyncio
import logging
from datetime import datetime

from pymodbus.datastore import (
    ModbusServerContext,
    ModbusSlaveContext,
    ModbusSparseDataBlock,
)

import server
import constants as c

update_path = c.UPDATE_PATH

r1_volt_path = c.R1_VOL_PATH
r1_volt_data = [c.TEST] * (c.CELL_COUNT + 1)


_logger = logging.getLogger(__name__)


async def updating_task(context):
    input_registers = 0x04
    holding_registers = 0x03

# incrementing loop
    while True:
        await asyncio.sleep(1)

        # setup time for first registers
        time_info = [datetime.now().year, datetime.now().month, datetime.now().day, datetime.now().hour,
                     datetime.now().minute, datetime.now().second]
        # Update time in Modbus
        context[0x01].setValues(holding_registers, 0x03, time_info)

        # setup system info
        system_info = [15, 16, 17, 18, 17, 16, 15]
        # update system info
        context[0x01].setValues(input_registers, 0x1007, system_info)

def setup_updating_server():
    """Run server setup."""
    network_info = ModbusSparseDataBlock(c.NETWORK_REGS, mutable=True)
    system_info = ModbusSparseDataBlock(c.SYSTEM_REGS, mutable=True)

    context = {
        0x01: ModbusSlaveContext(hr=network_info, ir=system_info)
    }
    context = ModbusServerContext(slaves=context, single=True)
    return server.setup_server(context=context)


async def run_updating_server(args):
    """Start updating_task concurrently with the current task."""
    task = asyncio.create_task(updating_task(args.context))
    task.set_name("Update Server")
    await server.run_async_server(args)  # start the server
    task.cancel()


async def main():
    """Combine setup and run."""
    run_args = setup_updating_server()
    await run_updating_server(run_args)


def get_in():
    asyncio.run(main(), debug=True)


get_in()


The script runs no issue on windows. connects with Modbus poll and updates the date in sl:1, reg:3, count 6. 
![image](https://github.com/user-attachments/assets/b5467d88-df3a-40ce-8e5d-9eeb46057e77)
not sure if that pic will show but it is Modbus poll on windows reading 3-8 with no errors.

The same script on Ubuntu 22.04 connects but gives a time out error
![image](https://github.com/user-attachments/assets/df167fa0-0691-4b59-a8c1-b30bac87b4a8)

I think this is more a Linux issue but I have not figured it out. using netstat -t, I can see the server established but just get time out.

Thank you again for any help you can provide.

**Reply by axgrinder** [reply]
More info. the green check shows its running. once Modbus poll or mbpoll is used, the exception is thrown.
![image](https://github.com/user-attachments/assets/59150672-2c23-4a1a-95fd-9ed21dc5408b)

**Reply by axgrinder** [reply]
was using 3.6.8, updated to 3.7.4 a got more errors
![image](https://github.com/user-attachments/assets/351dca19-4282-4f92-bccf-f06cd2bb0704)

**Reply by axgrinder** [reply]
However, 3.7.4 does still work on windows through pycharm.

**Reply by janiversen** [reply]
Updating from 3.6.8 to 3.7.4 requires changes in the app code !

the "KeyError: '127.0.0.1'" in the traceback suggest you are using 2 different versions of pymodbus !!!

Our CI runs on ubuntu and tests with python 3.9 ... 3.13 so I am sure StartAsyncServer works on Ubuntu / MacOS and Windows.

**Reply by janiversen** [reply, marked as answer]
please check with pip, with versions are installed on your windows and your ubuntu system.

the type error is because you have not specified framer= correctly, and exactly that have changed with 3.7.0
